src/AL_model.jl

```diff
@@ -0,0 +1,179 @@
+# based on Percival.jl
+# https://github.com/JuliaSmoothOptimizers/Percival.jl/blob/main/src/AugLagModel.jl
```

Comment:
What are the differences with Percival's AugLagModel, if any?

Reply:
> we could include also the composition of the nonsmooth term with a smooth one.

What do you mean exactly? We would bundle the smooth term with the other smooth terms in order to keep the prox operators as simple as possible, wouldn't we?

Reply:
Say the cost is of the form ``f(x)+g(F(x))`` for some smooth `f`, `F` and nonsmooth `g`. An option is to reformulate the problem with some additional variables to avoid the composition of `g` with `F`. This approach works but there is also another option, which eliminates the auxiliary variables. Applying the AL machinery gives an AL function with a term being the Moreau envelope of `g` composed with a shifted `F`, and one always evaluates only the simple prox operator of `g`.
My point is, the AL function depends on the problem formulation and there are several approaches when some `g` appears in the problem.